### ISSUE TITLE
Opus sound files?

### ISSUE THREAD

**Issue by kentonv** [gildor2/UModel#43]
I extracted a music file from Ark that claimed to be type "OPUS". However, it is not a standard Opus file -- it doesn't seem to use Ogg framing. Instead, it starts with the string "UE4OPUS", and then looks to be entirely binary from there on out.

Any ideas what this format might be / how to decode it? I imagine it's some alternative framing format for Opus packets so ought to be an easy translation, but I couldn't find any information on the internet about it.

I may be willing to contribute some code if I can figure this out.

(Note: I've never worked with the Unreal Engine, but I'm experienced with C++ and serialization formats.)

**Comment by gildor2** [comment]
There were some discussion about Opus sounds on the forum:
http://www.gildor.org/smf/index.php/topic,3317.msg23316.html#msg23316

UE4 rearranges Opus sound data for streaming, so formats doesn't match. What did you use to extract opus sounds from Ark?

I'm not familiar with Opus format by myself, so reading source code which performs this rearrangement is difficult for me, and I decided to not waste time trying to do its support.

**Comment by kentonv** [comment]
I used your tool to extract it from the uasset file, like:

    umodel -export -sounds GC_ARK_ElementChamber_Heavy.uasset

But this gave me a file `GC_ARK_ElementChamber_Heavy.OPUS` which is still non-standard.

Is the rearrangement code public? Any chance you can point me at it?

(FWIW I'm just trying to extract the music so I can listen to it. Not terribly important, but fun. :) )

**Comment by gildor2** [comment]
Ah I see, I've just found my own code which is doing that :) It simply dumps an array of data to a file, however in a case of streaming sounds they must be processed.

**Comment by kentonv** [comment]
Right. Do you know where I can find any further info about what processing is needed?

**Comment by gildor2** [comment]
Probably here:
Engine/Source/Developer/AudioFormatOpus/Private

**Comment by gildor2** [comment]
OPUS is used only for streamed sounds. And for instance, Fortnite uses pure ogg format for all sounds.

At the moment I'm not planning to spend any time on sounds formats, I have many other "next to do" things.

**Comment by bnnm** [comment]
I've added support for UE4OPUS in vgmstream (a library/plugins to play video game music formats, see here: https://github.com/kode54/vgmstream).

It works fine for ARK and Fortnite, but it only plays raw files extracted from the .uasset/ubulk/uexp. By "raw files" I mean the data starting from "UE4OPUS" string, de-chunked from ubulk+uexp (using the uexp table to remove blank padding) if needed. Here are a couple of examples: https://www.sendspace.com/file/t6o1n8

While I can extract them with some effort, I hope you can add raw .ue4opus extraction support since it'd be more future-proof and (I assume) easier for you. 
(at least I wasn't able to extract them with the current version, I always got an error).

**Comment by bnnm** [comment]
I could explain later how if you need it but it's a bit long, as you need to create valid Ogg pages which is kinda involved.

Even if you add a conversion option, I really, really hope you also add raw extraction, because I like to save the original UE4OPUS files, and avoid converted data whenever possible.
(IMO a UE4OPUS to OPUS converter would work better as an external program).

**Comment by gildor2** [comment]
I've added possibility to save streamed sounds in internal engine's format. For Fortnite, you'll get ".ue4opus" files as output. I didn't test results (I don't know how), but I think it should work.

Anyway, it would be much better if I'd be able to convert them into "standard" OPUS files.

**Comment by bnnm** [comment]
Thanks for adding the option.

***

As for converting UE4OPUS to Ogg Opus, here is a quick explanation. I may add more detail later but Ogg is stupidly complex and it's very time consuming.

If you add this conversion I hope you still keep the option to extract it raw, though.



UE4OPUS starts with a simple header: string, sample rate, total samples, channels (nothing in the UE4 classes is needed). Then Opus packets (packet size + packet data) until file end.

You need to take the header info and recreate a OggS "opus header" page, make a fake OggS "comment" page, then put Opus packets inside many OggS "data" pages.

Ogg is divided into "OggS pages", that can contain certain fixed fields (page type, position/sequence, stream id, crc) and a variable "lacing values" table (that has the number of packets and their sizes, in a weird format).

Inside a "data page" you must fill the lacing table and put N opus packets inside (without UE4OPUS packet size). N can be any number but not too much nor too little, as it'd be inefficient.

Ogg pages need position, which means number of samples up to that point, you must manually extract and count samples in opus packets (this can be found in the opus src), though maybe all packets share the same number of samples (not sure). Ogg pages are also CRC-protected.

Note that in the "opus header" you should also put 'encoder delay', an extra field not in UE4OPUS that is equal to (samples in first packet / 8).

Anyway, getting any one field wrong may result in an invalid Ogg, so as you can see it's all quite complex.

The Ogg page formats are explained in Xiph's docs, though it's a bit hard to follow, you may have to check their sources.

**Comment by gildor2** [comment]
Thank you for long and very detailed explanation. Hopefully I'll return to "true OPUS" support one day. However, a question: what's the point for supporting "UE4OPUS" if I'll support "OPUS"?

**Comment by bnnm** [comment]
Main reason is data preservation, I prefer to store original files only.

If I have the original data, I can play it as-is (w/ vgmstream) or convert to Ogg Opus if I wanted. As you can see Ogg is rather complex, so if umodel somehow gets the Ogg conversion wrong one could extract the original UE4OPUS and check it.

Or if Epic changes the format or adds features that Ogg Opus can't use (like looping, common in videogames formats) having the original files around would be useful for me.

**Comment by gildor2** [comment]
Shouldn't "UE4OPUS" differ from Ogg or OPUS with just data layout? I feel you afraid of transcode data with losing quality.

**Comment by bnnm** [comment]
I understand data quality remains the same. But as a principle, and for various tiny details, unmodified data is only one I can trust to play 100% like the game.

Generally, most "format to other format" converters don't understand videogame audio needs (like looping, encoder delay skip, proper sample counts, non-standard sample rates, seek tables, etc) and end up destroying valuable (to me) metadata. Some just want playable audio but I look for 100% accurate audio, so that metadata is worth keeping.

For example, if don't make the Ogg Opus *very* carefully, you may end up with a file +-0.2 secs longer than the original UE4OPUS. Which may seem small, but for accurate looping/gapless sound it's enough to feel off.

Anyway, it's your tool and you can do what you feel right. Just remember some appreciate the option to keep original data. :)

**Comment by gildor2** [comment]
Sure :)
